src/DataSource.ts

```diff
@@ -69,10 +69,15 @@ export class DataSource extends DataSourceApi<GrafanaQuery, GenericOptions> {
   }
 
   metricFindQuery(query: string, options?: any, type?: string): Promise<MetricFindValue[]> {
-    const interpolated = {
-      type,
-      target: getTemplateSrv().replace(query, undefined, 'regex'),
-    };
+    let interpolated;
+    try {
+      interpolated = JSON.parse(getTemplateSrv().replace(query, undefined, 'json'));
+    } catch (e) {
```

Comment:
we should check for specific error here, not catch all

Reply:
Yeah I was thinking about how best to deal with this, because it's not really to catch an error so much as it is a check to see if the inputted query string is actual JSON or not (because we still want to be backwards compatible with people who want to send the legacy `{"target":"querystringhere"}` format I assume). I looked it up and apparently this really is the accepted way of checking if a string is valid JSON. I didn't love the idea of a try/catch myself either though so I understand your initial hesitation there.

It seems for the template variable query strings, we don't have the ability to modify the entry field to add something like a checkbox or anything resembling the "Additional JSON data" section like custom queries (which can come with their own query editor for that section that we can assume is JSON). Otherwise I would say we could put a checkbox there for the user to say, "This is a json string" and handle it explicitly rather than inferring based on the JSON validity of the string.

We could consider adding an option in the datasource configuration to say, "I use JSON in my template query strings" and then we could look at that setting in the function before we try to parse the string as JSON. Would you prefer that? The downside to that would be that it's one or the other; you're either using JSON in the template queries or you're using the traditional target/string approach. And then you're sort of at the whim of the person who configured the datasource to choose that for everyone who might Editors, but that's probably not a big concern considering the different queries have to be handled on the datasource backend anyway.

I digress, to me, the way I've implemented it here is the easiest and I do sort of feel like the try/catch fits the spirit of the plugin: If you are using JSON, we _try_ to send it to the backend datasource as that JSON. Otherwise, we _catch_ that you are not using valid JSON, and we force it into JSON by plopping it into `{"target":"yourquerystring"}`.

What do you think? Either way, we have to check for JSON or let the user tell us that it is JSON, or make a breaking change by telling everyone that they _must_ provide JSON here.

Reply:
Oh my mistake, there is a [VariableQueryEditor](https://grafana.com/docs/grafana/latest/developers/plugins/add-support-for-variables/#add-support-for-query-variables-to-your-data-source). Ok let me take a look at how that works and maybe we can make the option to use JSON or not more explicit.

Reply:
Ok I was able to get a first draft of the toggle switch working in a new VariableQueryEditor component!

![image](https://user-images.githubusercontent.com/529833/111931488-ad75c680-8a91-11eb-9383-e69bc4083287.png)

There is still an issue with toggling it on/off and I'm not sure about the new type yet, so let me try all the other features and see about making tests.